### PR TITLE
rpc: Drop migratewallet experimental warning

### PR DESCRIPTION
**Pull request by achow101**
The migration process itself hasn't fundamentally changed since it was added, so I think it's reasonable to say that it is no longer experimental.

### CODE REVIEW

**Review by furszy — COMMENTED** [review]
There is a small bug in the addressbook migration code (fixed by #26836) that we might want to fix before removing the experimental warning.

Probably, I could decouple the fix and test coverage into an standalone PR so we can get more attention there.

**Review by furszy — COMMENTED** [review]
> Would it make sense to document that this is expected to take a long time on spinning storage? For me a freshly created BDB wallet takes ~5 minutes to migrate.

Created #28574 to improve the situation. Let me know how it goes with it.

**Comment by willcl-ark** [comment]
Is the plan here to review #28574 and try and get that in, if it improves things, or should we remove the warning as proposed here and add a note that it may take more time?

I'd be happy to review/test either when I know which approach we'd prefer.

**Review by furszy — COMMENTED** [review]
> Is the plan here to review #28574 and try and get that in, if it improves things, or should we remove the warning as proposed here and add a note that it may take more time?
> 
> I'd be happy to review/test either when I know which approach we'd prefer.

I would rather leave this warning for another release just so the process undergoes an entire release cycle without issues prior to dropping the "experimental" label. I don't think that keeping it for another cycle would modify the legacy wallet removal schedule at all.

The reality is that, even though they may seem harmless, we continue encountering issues with this process.. e.g. #28609 and #28610. And I think that we should try to avoid, as much as possible, any unexpected behavior resulting from the wallet migration process as it could be quite alarming for our users.

Other than that, after reviews on #28609 and #28610, I would be happy to receive you on #28574. A process which its execution takes +5 minutes on a spinning disk hard drive seems pretty impractical for me (and could also be tagged as an issue).

**Comment by achow101** [comment]
I've pushed a commit that mentions the time in the RPC doc.

**Comment by stickies-v** [comment]
> We have the legacy wallet scheduled to be removed for 27.0

According to #20160, it seems we're only stopping creation of new wallets in 27.0, and removing BDB in 28.0? Given that there still are outstanding bugs (even if not necessarily super critical), removing the experimental label for 26.0 is unnecessarily risky in my view

**Comment by achow101** [comment]
Moved to the 27.0 milestone.

File: src/wallet/rpc/wallet.cpp [hunk 1]
```diff
@@ -739,13 +739,13 @@ RPCHelpMan simulaterawtransaction()
...
     return RPCHelpMan{"migratewallet",
-        "EXPERIMENTAL warning: This call may not work as expected and may be changed in future releases\n"
         "\nMigrate the wallet to a descriptor wallet.\n"
         "A new wallet backup will need to be made.\n"
         "\nThe migration process will create a backup of the wallet before migrating. This backup\n"
         "file will be named <wallet name>-<timestamp>.legacy.bak and can be found in the directory\n"
         "for this wallet. In the event of an incorrect migration, the backup can be restored using restorewallet."
-        "\nEncrypted wallets must have the passphrase provided as an argument to this call.",
+        "\nEncrypted wallets must have the passphrase provided as an argument to this call.\n"
+        "\nThis RPC may take a long time to complete. Increasing the RPC client timeout is recommended.",
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/rpc/wallet.cpp:748]: Should we then include #26008 and maybe #28574 in 27.0?

**Comment by murchandamus** [comment]
Hey, I was taking a look at this PR. What is the status on this? Are we aiming to get this into 27.0? What about the two PRs, that @S3RK mentioned, they are open and in draft. Does this PR here depend on them?

**Review by furszy — COMMENTED** [review]
> Hey, I was taking a look at this PR. What is the status on this? Are we aiming to get this into 27.0? What about the two PRs, that @S3RK mentioned, they are open and in draft. Does this PR here depend on them?

#28574 is in draft because it is the parent PR. It consolidates the entire work in a single location, allowing tests and benchmarks to be performed against the end goal as well.
This work was decoupled into three independent PRs: #28894 (merged), #28987, #26836. There is no specific order required for the review.

I think we should aim to get this one in for v27. The process shouldn't take ~5 minutes to complete (https://github.com/bitcoin/bitcoin/pull/28037#issuecomment-1628507151).  Since there is no feedback during the process, users might end up aborting it, thinking it has stalled, and leave their wallet in an inconsistent state, requiring manual recovery from the backup.

**Comment by furszy** [comment]
Also, we do have other bug fixes that should be included in v27 to deprecate this warning.
See #28976, #28868, #29112. Would be nice to bring more attention to them.

**Review by josibake — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/28037/commits/f1684bb88a878eb3f54e945db39ea69b14256eef

Looks like all the known bugs have been fixed and most of the performance improvements have been merged. The only two outstanding performance improvements are #28987 and #26008, both of which seem close/RFM. Given that we are also including a warning that this process is expected to take longer, I think its safe to drop the warning even without #28987 and #26008 (obviously, better if we can also get those in with v27).

**Review by furszy — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/commit/f1684bb88a878eb3f54e945db39ea69b14256eef

> The only two outstanding performance improvements are https://github.com/bitcoin/bitcoin/pull/28987 and https://github.com/bitcoin/bitcoin/pull/26008

There are actually more. #28987, and also #29403, are prerequisites for further performance improvements on #28574.

That being said, while I would like to reach that goal (same as getting closer in #26008), I also think that we should move forward with this one.

**Comment by achow101** [comment]
> I tested yesterday a bdb testnet3 wallet with 1000 txs on spinning storage, and I aborted the migration after 20 minutes.

Can you profile that so we can see where it is actually spending the most time?

**Review by furszy — COMMENTED** [review]
The last batching changes might help but.. it smells like you are facing another issue apart from those. Profiling results would help diagnose the problem.

**Comment by achow101** [comment]
@maflcko Can you provide any more details about the wallet? The output of `getwalletinfo` would be helpful.

**Comment by achow101** [comment]
How long does it take to migrate if you just let it complete?

**Review by furszy — COMMENTED** [review]
Is the chain sync, reindex or any indexing procedure taking that long as well?

**Review by furszy — COMMENTED** [review]
> Not sure. I am using the spinning storage only to test migratewallet.

I'm asking because if other processes are equally slow, then this could be "ok" (and we could declare the hardware unsupported for core entirely...). But if only this process takes forever, then we are sure that there is an issue.

> It finished after 2 hours.

We are definitely not expecting this process to take that long. The wallet you shared is really small.

> Another problem is that (I presume) large single-keys-bag wallets may take hundred times longer to load after migration, even on fast storage, IIRC. If this is still the case, it seems like a higher priority to fix, because this is a cost to be paid every time the wallet is loaded.
> I guess time loadwallet will be improved by https://github.com/bitcoin/bitcoin/pull/26008#issuecomment-1943469180

It will, but #26008's current implementation comes with higher memory costs, which is not that good if your wallet is running on a resource-constrained system. Yet, (personal view) it will be nice to tackle one issue at a time. Users need to be able to migrate their wallet in a decent amount of time before reaching the point of running a big single-key-bag wallet.

-------------

Continuing with the issue investigation, could you measure how much it takes to create a fresh wallet? It would be good to know how much it takes to top up the new key pool for the 8k default keys (the same process is executed during migration).

Also, how much available memory (ram) does your system have?.

And, just in case, try running #28574 too.

**Comment by furszy** [comment]
Update: I'm cooking another set of improvements.. will try to have them for tonight.

**Comment by furszy** [comment]
Or well, I just crafted a simple version of what I'm doing (because I'm planning to create a statement builder around this work, and that will take me some more time).

Could you please try this branch: https://github.com/furszy/bitcoin-core/tree/2024_wallet_batch_migration_multi_insert and share the results?

It is essentially #28574 + 506b73872b9ce232e8c6b9fefbdb0c33e04d0707. The new commit introduces a raw multi-insert statement. Locally, this shows a significant speedup but I'm on a SSD.

**Comment by willcl-ark** [comment]
I tried to reproduce using a wallet with ~1k tx located on a spinning disk -- although this time on regtest -- but I found `migratewallet` to be pleasantly fast:

```fish
will@will-ubuntu in ~/src/bitcoin on  master [$?⇡] : C v17.0.2-clang : 🐍 3.9.18
₿ time /home/will/src/bitcoin/src/bitcoin-cli -regtest -datadir=/media/will/PEPSI/big-wallet-900-tx-original migratewallet big
{
  "wallet_name": "big",
  "backup_path": "/media/will/PEPSI/big-wallet-900-tx-original/regtest/wallets/big/big-1707927351.legacy.bak"
}

________________________________________________________
Executed in    8.14 secs      fish           external
   usr time    1.69 millis  226.00 micros    1.46 millis
   sys time    0.45 millis   45.00 micros    0.40 millis


will@will-ubuntu in ~/src/bitcoin on  master [$?⇡] : C v17.0.2-clang : 🐍 3.9.18
₿ /home/will/src/bitcoin/src/bitcoin-cli -regtest -datadir=/media/will/PEPSI/big-wallet-900-tx-original getwalletinfo
{
  "walletname": "big",
  "walletversion": 169900,
  "format": "sqlite",
  "balance": 13606.24414485,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 515.63072317,
  "txcount": 952,
  "keypoolsize": 4000,
  "keypoolsize_hd_internal": 4000,
  "paytxfee": 0.00000000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false,
  "descriptors": true,
  "external_signer": false,
  "blank": false,
  "birthtime": 0,
  "lastprocessedblock": {
    "hash": "16632672917b708d311b8197d7be852c9a5ce41413b04ecfc2ca72637148766b",
    "height": 634
  }
}
```

The BDB wallet `getwalletinfo` (slightly smaller keypool size):

```fish
will@will-ubuntu in ~/src/bitcoin on  master [$?⇡] : C v17.0.2-clang : 🐍 3.9.18
₿ /home/will/src/bitcoin/src/bitcoin-cli -regtest -datadir=/media/will/PEPSI/big-wallet-900-tx-original getwalletinfo
{
  "walletname": "/home/will/bitcoin-chains/big-wallet-900-tx-original/regtest/wallets/big",
  "walletversion": 169900,
  "format": "bdb",
  "balance": 13606.24414485,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 515.63072317,
  "txcount": 952,
  "keypoololdest": 1707925846,
  "keypoolsize": 1000,
  "hdseedid": "f48cadf402473b7726449f7e50c836466970299e",
  "keypoolsize_hd_internal": 1000,
  "paytxfee": 0.00000000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false,
  "descriptors": false,
  "external_signer": false,
  "blank": false,
  "birthtime": 1707925845,
  "lastprocessedblock": {
    "hash": "16632672917b708d311b8197d7be852c9a5ce41413b04ecfc2ca72637148766b",
    "height": 634
  }
}
```

**Comment by willcl-ark** [comment]
I was able to reproduce it with wallet version 60000 and 1k keys/tx I also see 30 mins...

```fish
₿ /home/will/src/bitcoin/src/bitcoin-cli -regtest -datadir=/media/will/PEPSI/regtest-v60000 getwalletinfo
{
  "walletname": "",
  "walletversion": 60000,
  "format": "bdb",
  "balance": 14716.40625000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 78.12500000,
  "txcount": 1000,
  "keypoololdest": 1707941916,
  "keypoolsize": 1000,
  "paytxfee": 0.00000000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false,
  "descriptors": false,
  "external_signer": false,
  "blank": false,
  "birthtime": 1707941702,
  "lastprocessedblock": {
    "hash": "369d42451890f636c521f3d3d70241aa524eb36cfa53c4cd16248d3b9f772538",
    "height": 1000
  }
}

will@will-ubuntu in ~ : 🐍 3.11.6
₿ time /home/will/src/bitcoin/src/bitcoin-cli -regtest -datadir=/media/will/PEPSI/regtest-v60000 -rpcclienttimeout=9999999999 migratewallet ""
{
  "wallet_name": "",
  "backup_path": "/media/will/PEPSI/regtest-v60000/regtest/-1707945979.legacy.bak"
}

________________________________________________________
Executed in   32.20 mins      fish           external
   usr time    2.44 millis    0.00 micros    2.44 millis
   sys time    0.37 millis  374.00 micros    0.00 millis

```

I was trying to get a flamegraph but something seems to have broken with my `perf` tool after upgrading kernels...

**Comment by furszy** [comment]
q: the wallet to be migrated, is already loaded by the time you call `migratewallet`? Or you are migrating a non-loaded wallet?

Because, if the wallet wasn't loaded by the newer version yet, it will run the `UpgradeKeyMetadata` process during migration too. Which involves performing writes for 2 times the number of your key pool.

I just updated my branch batching that process too. In case someone want to test: https://github.com/furszy/bitcoin-core/tree/2024_wallet_batch_migration_multi_insert.
Will try to get an hdd tomorrow if this is not solved by then.

**Comment by achow101** [comment]
I'm not entirely convinced that migration is actually spending most of the time in disk I/O for these big wallets.

This is a flamegraph of the migration of my big nonhd wallet on master:
master-migrate-big-nonhd-perf

Ignoring the locking stuff since I didn't turn off the lock debugging, it seems like a big chunk of time is spent in `IsMine` after the descriptors have been created. This suggests to me that #26008 would provide a significant speedup there.

Here's a flamegraph of migration of the same wallet with #26008:
ismine-cache-migrate-big-nonhd-perf

You can see that almost no time is spent in `IsMine`, so I think #26008 should make migration of such wallets much faster:

```
$ src/bitcoin-cli -regtest -rpcwallet=/run/media/ava/F5CD-B8E0/big_nonhd_wallet migratewallet
{
  "wallet_name": "/run/media/ava/F5CD-B8E0/big_nonhd_wallet",
  "backup_path": "/run/media/ava/F5CD-B8E0/big_nonhd_wallet-1707974252.legacy.bak"
}

________________________________________________________
Executed in  150.25 secs      fish           external
   usr time    4.69 millis    0.00 micros    4.69 millis
   sys time    0.41 millis  414.00 micros    0.00 millis
```

If I turn off debugging, migration of my test wallet is also way faster:
```
$ src/bitcoin-cli -regtest -rpcwallet=/run/media/ava/F5CD-B8E0/big_nonhd_wallet migratewallet
{
  "wallet_name": "/run/media/ava/F5CD-B8E0/big_nonhd_wallet",
  "backup_path": "/run/media/ava/F5CD-B8E0/big_nonhd_wallet-1707974585.legacy.bak"
}

________________________________________________________
Executed in   82.31 secs      fish           external
   usr time    0.00 millis    0.00 micros    0.00 millis
   sys time    4.33 millis  407.00 micros    3.92 millis
```

@maflcko @willcl-ark Can you say how many keys (or descriptors after migration) the wallet has? Can you also try migrating it using #26008?

**Comment by achow101** [comment]
I'm any case, I didn't think the performance issues should be a blocker for this PR. It's clear that migration works, it may just take a while.

Missing this release _will_ push back the timeline another release cycle, if not more.

**Review by ryanofsky — APPROVED** [review]
Code review ACK f1684bb88a878eb3f54e945db39ea69b14256eef

"This RPC may take a long time to complete" seems like much more helpful and accurate documentation than "This call may not work as expected and may be changed in future releases" at this point.

**Review by willcl-ark — APPROVED** [review]
ACK f1684bb88a878eb3f54e945db39ea69b14256eef

Agree that this should no longer be marked as experimental in the next release even though some migrations do appear to take a(n unexpectedly) long time.

**Comment by willcl-ark** [comment]
@achow101 for context my test wallet has 8016 descriptors after migration:

```fish
will@will-ubuntu in ~ : 🐍 3.11.6
₿ /home/will/src/bitcoin/src/bitcoin-cli -regtest -datadir=/media/will/PEPSI/regtest-v60000 listdescriptors | jq '.descriptors | length'
8016
```

I have also in the meantime tested a branch from @furszy which reduced migration time from 30 mins to 80 seconds 👀

**Comment by furszy** [comment]
> I have also in the meantime tested a branch from @furszy which reduced migration time from 30 mins to 80 seconds 👀

That's great!
@maflcko could you also try it please?.

Branch: https://github.com/furszy/bitcoin-core/tree/2024_wallet_batch_migration_multi_insert.

**Pull request merged**